File: pyracms-cpp-port/cli_pkg/sigv4.py
"""AWS Signature Version 4 for S3 (stdlib only): header auth + presign."""
import hashlib
import hmac
import time
from urllib.parse import quote, unquote

EMPTY_SHA = hashlib.sha256(b"").hexdigest()
ALGO = "AWS4-HMAC-SHA256"


def uri_encode(s, keep_slash=False):
    return quote(s, safe="/-_.~" if keep_slash else "-_.~")


def canonical_query(raw):
    """Decode, encode once, sort by name then value; bare key -> key=."""
    pairs = []
    for part in filter(None, raw.split("&")):
        k, _, v = part.partition("=")
        pairs.append((uri_encode(unquote(k)), uri_encode(unquote(v))))
    return "&".join(f"{k}={v}" for k, v in sorted(pairs))


def amz_now():
    return time.strftime("%Y%m%dT%H%M%SZ", time.gmtime())


def _hmac(key, msg):
    return hmac.new(key, msg.encode(), hashlib.sha256).digest()


def sign(key, method, path, query, headers, payload_sha, amzdate,
         presigned=False):
    """Returns (signature, authorization). `headers`: lower-case names
    (must hold host); x-amz-date/x-amz-content-sha256 are added unless
    presigned. `key` = (access, secret, region)."""
    access, secret, region = key
    h = dict(headers)
    if not presigned:
        h["x-amz-date"] = amzdate
        h["x-amz-content-sha256"] = payload_sha
    names = sorted(h)
    canon_h = "".join(f"{n}:{h[n].strip()}\n" for n in names)
    signed = ";".join(names)
    canon = "\n".join([method, uri_encode(unquote(path), True),
                       canonical_query(query), canon_h, signed,
                       payload_sha])
    day = amzdate[:8]
    scope = f"{day}/{region}/s3/aws4_request"
    to_sign = "\n".join([ALGO, amzdate, scope,
                         hashlib.sha256(canon.encode()).hexdigest()])
    k = _hmac(("AWS4" + secret).encode(), day)
    for part in (region, "s3", "aws4_request"):
        k = _hmac(k, part)
    sig = hmac.new(k, to_sign.encode(), hashlib.sha256).hexdigest()
    auth = (f"{ALGO} Credential={access}/{scope}, "
            f"SignedHeaders={signed}, Signature={sig}")
    return sig, auth

File: pyracms-cpp-port/backend/src/storage/SigV4.cpp
#include "storage/SigV4.h"

namespace pyracms {

static std::string trim(const std::string &v) {
    auto b = v.find_first_not_of(" \t");
    auto e = v.find_last_not_of(" \t");
    return b == std::string::npos ? "" : v.substr(b, e - b + 1);
}

std::string sigv4Sign(const SigV4Key &key, const SigV4Request &req,
                      std::string *authorization) {
    auto hdrs = req.headers;
    if (!req.presigned) {
        hdrs["x-amz-date"] = req.amzDate;
        hdrs["x-amz-content-sha256"] = req.payloadSha;
    }
    std::string canonHeaders, signed_;
    for (auto &h : hdrs) { // std::map: already sorted by name
        canonHeaders += h.first + ":" + trim(h.second) + "\n";
        signed_ += (signed_.empty() ? "" : ";") + h.first;
    }
    // S3: the path is encoded exactly once (no double encoding).
    auto canon = req.method + "\n" +
                 sigv4UriEncode(sigv4DecodePath(req.path), true) + "\n" +
                 sigv4CanonicalQuery(req.query) + "\n" + canonHeaders +
                 "\n" + signed_ + "\n" + req.payloadSha;
    auto day = req.amzDate.substr(0, 8);
    auto scope = day + "/" + key.region + "/s3/aws4_request";
    auto toSign = "AWS4-HMAC-SHA256\n" + req.amzDate + "\n" + scope +
                  "\n" + sha256Hex(canon);
    auto k = hmacSha256("AWS4" + key.secret, day);
    k = hmacSha256(k, key.region);
    k = hmacSha256(k, "s3");
    k = hmacSha256(k, "aws4_request");
    auto sig = toHex(hmacSha256(k, toSign));
    if (authorization)
        *authorization = "AWS4-HMAC-SHA256 Credential=" + key.access + "/" +
                         scope + ", SignedHeaders=" + signed_ +
                         ", Signature=" + sig;
    return sig;
}

} // namespace pyracms
